mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -0,0 +1,931 @@
+			CheckStatusAndThrow (result);
+			return new List<SslCipherSuite> (ciphers);
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetNumberEnabledCiphers (/* SSLContextRef */ IntPtr context, /* size_t* */ out nint numCiphers);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetEnabledCiphers (/* SSLContextRef */ IntPtr context, SslCipherSuite *ciphers, /* size_t* */ ref nint numCiphers);
+
+		public unsafe IList<SslCipherSuite> GetEnabledCiphers ()
+		{
+			nint n;
+			var result = SSLGetNumberEnabledCiphers (Handle, out n);
+			CheckStatusAndThrow (result);
+			if ((result != SslStatus.Success) || ((int)n <= 0))
+				return null;
+
+			var ciphers = new SslCipherSuite [(int)n];
+			fixed (SslCipherSuite *p = ciphers) {
+				result = SSLGetEnabledCiphers (Handle, p, ref n);
+			}
+			CheckStatusAndThrow (result);
+			return new List<SslCipherSuite> (ciphers);
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLSetEnabledCiphers (/* SSLContextRef */ IntPtr context, SslCipherSuite *ciphers, /* size_t */ nint numCiphers);
+
+		public unsafe void SetEnabledCiphers (SslCipherSuite [] ciphers)
+		{
+			if (ciphers == null)
+				throw new ArgumentNullException ("ciphers");
+
+			SslStatus result;
+
+			fixed (SslCipherSuite *p = ciphers)
+				result = SSLSetEnabledCiphers (Handle, p, (IntPtr)ciphers.Length);
+			CheckStatusAndThrow (result);
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetNegotiatedCipher (/* SSLContextRef */ IntPtr context, /* SslCipherSuite* */ out SslCipherSuite cipherSuite);
+
+		public SslCipherSuite NegotiatedCipher {
+			get {
+				SslCipherSuite value;
+				var result = SSLGetNegotiatedCipher (Handle, out value);
+				CheckStatusAndThrow (result);
+				return value;
+			}
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetPeerDomainNameLength (/* SSLContextRef */ IntPtr context, /* size_t* */ out nint peerNameLen);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetPeerDomainName (/* SSLContextRef */ IntPtr context, /* char* */ byte[] peerName, /* size_t */ ref nint peerNameLen);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLSetPeerDomainName (/* SSLContextRef */ IntPtr context, /* char* */ byte[] peerName, /* size_t */ nint peerNameLen);
+
+		public string PeerDomainName {
+			get {
+				nint length;
+				var result = SSLGetPeerDomainNameLength (Handle, out length);
+				CheckStatusAndThrow (result);
+				if (result != SslStatus.Success || (int)length == 0)
```

Comment:
Be very careful with that - it is **not** used 6 times in the block if you look closely at the code.

It is used twice, then `SSLGetPeerDomainName()` calls it in a `ref` context - which may modify the value - then it's used four more times.